### PR TITLE
[Playwright] [PEPPER-293] Version upgrade to latest

### PR DESCRIPTION
**Pull request by aweng98**
Upgrade Playwright to the latest version. Verified tests are still working after upgrade.

- Running tests on localhost:

Screen Shot 2022-12-01 at 10 15 43 AM

- Running tests inside docker container:

Screen Shot 2022-12-01 at 1 03 56 PM

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ executors:
     resource_class: medium+
     docker:
       # Playwright docker image version MUST match Playwright version in project
-      - image: mcr.microsoft.com/playwright:v1.25.0-focal
+      - image: mcr.microsoft.com/playwright:v1.28.0-focal
```
Comment (aweng98) [inline .circleci/config.yml:40]: https://playwright.dev/docs/docker#pull-the-image

File: playwright-e2e/pages/page-base.ts [hunk 2]
```diff
@@ -73,8 +73,20 @@ export default abstract class PageBase implements PageInterface {
...
+   *  <br> - forceClick: Whether to bypass error message checks before click. If set to true, click regardless an error on page. Defaults to false.
+   *  <br> - waitForNav: Whether to wait for page navigation to complete after click. Defaults to false.
+   * @returns {Promise<void>}
+   * @protected
+   */
+  protected async clickAndWaitForNav(locator: Locator, opts: { waitForNav?: boolean; forceClick?: boolean } = {}): Promise<void> {
+    const { waitForNav = false, forceClick = false } = opts;
+    if (!forceClick) {
+      await expect(this.page.locator('.error-message')).toBeHidden();
+    }
```
Comment (aweng98) [inline playwright-e2e/pages/page-base.ts:89]: Add new parameter `forceClick` to address one flaky failure I saw during testing. An error message did not go away for 10 seconds, so the click a button has failed. I don't know why it happened.

File: playwright-e2e/playwright.config.ts [hunk 3]
```diff
@@ -46,7 +46,7 @@ const testConfig: PlaywrightTestConfig = {
 
   /* Reporter to use. See https://playwright.dev/docs/test-reporters */
   reporter: [
-    ['html', { open: 'never', outputFolder: 'html-test-results' }],
+    ['html', { open: 'never', outputFolder: 'html-test-results', host: '0.0.0.0', port: 9223 }],
```
Comment (aweng98) [inline playwright-e2e/playwright.config.ts:49]: Allow html test result inside docker container to be opened outside docker container.

File: playwright-e2e/lib/widget/dropdown.ts [hunk 4]
```diff
@@ -6,21 +6,30 @@ import { Locator, Page } from '@playwright/test';
...
+  private readonly label: string;
 
-  constructor(page: Page, label: string | RegExp) {
+  constructor(page: Page, label: string, opts: { exactMatch?: boolean } = {}) {
+    const { exactMatch = true } = opts;
     this.page = page;
-    this.locator = this.page.locator('li.dropdown').filter({ has: this.page.locator('a.dropdown-toggle'), hasText: label });
+    this.label = label;
+    const textSelector = exactMatch ? `text="${this.label}"` : `text=${this.label}`;
+    this.locator = this.page.locator('li.dropdown').filter({ has: this.page.locator(textSelector) });
```
Comment (aweng98) [inline playwright-e2e/lib/widget/dropdown.ts:16]: Old locator no longer works in new Playwright version. 😖

File: playwright-e2e/README.md [hunk 5]
```diff
@@ -49,20 +49,23 @@ Note: Update docker image version when upgrading Playwright version
 - Change dir to **/playwright-e2e**
   > cd playwright-e2e/
 
-- Start Playwright docker image
-  > docker run -v $PWD:/e2e -w /e2e -it --rm --ipc=host mcr.microsoft.com/playwright:v1.25.0-focal /bin/bash
+- Start running Playwright docker image
+  > docker run -v $PWD:/e2e -w /e2e -it --rm -d --ipc=host -p 9323:9323 mcr.microsoft.com/playwright:v1.28.0-focal /bin/bash
```
Comment (aweng98) [inline playwright-e2e/README.md:53]: Allow html test result inside docker container to be opened outside docker container.

File: playwright-e2e/lib/widget/dropdown.ts [hunk 6]
```diff
@@ -6,21 +6,30 @@ import { Locator, Page } from '@playwright/test';
...
+    return this.label;
+  }
+
   /**
    * Determine if dropdown is open by look up aria-expanded property
    */
   async isOpen(): Promise<boolean> {
-    return (await this.toLocator().locator('a.dropdown-toggle').getAttribute('aria-expanded')) === 'true';
+    const ariaExpanded = await this.toLocator().locator('a[data-toggle="dropdown"]').getAttribute('aria-expanded');
+    return ariaExpanded ? Boolean(ariaExpanded) : false;
```
Comment (aweng98) [inline playwright-e2e/lib/widget/dropdown.ts:26]: No longer works in new version. Updated to handle Null pointer explicitly.

**Comment by aweng98** [comment]
@Trimax @bskinner PTAL.

File: playwright-e2e/lib/widget/dropdown.ts [hunk 7]
```diff
@@ -7,9 +7,11 @@ export default class Dropdown {
   private readonly page: Page;
   private readonly locator: Locator;
 
-  constructor(page: Page, label: string | RegExp) {
+  constructor(page: Page, label: string, opts: { exactMatch?: boolean } = {}) {
+    const { exactMatch = true } = opts;
     this.page = page;
-    this.locator = this.page.locator('li.dropdown').filter({ has: this.page.locator('a.dropdown-toggle'), hasText: label });
+    const textSelector = exactMatch ? `text="${label}"` : `text=${label}`;
+    this.locator = this.page.locator('li.dropdown').filter({ has: this.page.locator(textSelector) });
```
Comment (aweng98) [inline playwright-e2e/lib/widget/dropdown.ts:14]: Old locator no longer works in new Playwright version. 😖

**Review by Trimax — APPROVED** [review]
Approving the PR, but TypeScript is not my area of expertise. I see it for the 1st time in my life.

**Review by bskinner — APPROVED** [review]
The PR looks reasonable to me 👍

**Comment by aweng98** [comment]
Thank you all.

**Pull request merged**